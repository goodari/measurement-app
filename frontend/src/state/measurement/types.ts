export interface Measurement {
  date: string;
  value: number;
}

export interface MeasurementState {
  measurements: Measurement[];
}
