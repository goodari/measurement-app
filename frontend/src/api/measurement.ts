import { axiosInstance } from "./axiosInstance";

const MEASUREMENT_URL = "/measurements";

export const postMeasurement = (value: number) =>
  axiosInstance.post(`${MEASUREMENT_URL}`, { value });
