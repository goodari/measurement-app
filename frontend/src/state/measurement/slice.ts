import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { getMeasurements, postMeasurement } from "../../api/measurement";
import { MeasurementState } from "./types";

export interface CounterState {
  value: number;
}

const initialState: MeasurementState = {
  measurements: [],
};

const createMeasurement = createAsyncThunk(
  "measurements/create",
  async (value: number) => {
    const response = await postMeasurement(value);

    return response.data;
  }
);

const fetchMeasurements = createAsyncThunk("measurements/fetch", async () => {
  const response = await getMeasurements();

  return response.data;
});

export const measurementSlice = createSlice({
  name: "measurement",
  initialState,
  reducers: {},
  extraReducers: (builder) =>
    builder
      .addCase(createMeasurement.fulfilled, (state, action) => {
        state.measurements.push(action.payload);
      })
      .addCase(fetchMeasurements.fulfilled, (state, action) => {
        state.measurements = action.payload.measurements;
      }),
});

export { createMeasurement, fetchMeasurements };

export default measurementSlice.reducer;
