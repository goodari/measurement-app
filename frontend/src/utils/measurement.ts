import { Measurement } from "../state/measurement/types";

export enum Order {
  ASC = -1,
  DESC = 1,
}

export const sortByDate = (order: Order = Order.ASC) => (
  a: Measurement,
  b: Measurement
) =>
  (order === Order.ASC ? -1 : 1) *
  (new Date(b.date).getTime() - new Date(a.date).getTime());

export const sortByValue = (order: Order = Order.ASC) => (
  a: Measurement,
  b: Measurement
) => (order === Order.ASC ? -1 : 1) * (a.value < b.value ? 1 : -1);
