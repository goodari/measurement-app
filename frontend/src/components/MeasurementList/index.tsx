import React, { FunctionComponent, useEffect, useState } from "react";
import { Measurement } from "../../state/measurement/types";
import { Order, sortByDate, sortByValue } from "../../utils/measurement";

type PropType = {
  measurements: Measurement[];
};

enum MeasurementColumn {
  Date,
  Value,
}

type Col = {
  id: MeasurementColumn;
  label: string;
  sortFunction: (order: Order) => any;
};

const columns: Col[] = [
  {
    id: MeasurementColumn.Date,
    label: "Date",
    sortFunction: sortByDate,
  },
  {
    id: MeasurementColumn.Value,
    label: "Value",
    sortFunction: sortByValue,
  },
];

const MeasurementList: FunctionComponent<PropType> = ({ measurements }) => {
  const [sortedBy, setSortedBy] = useState(MeasurementColumn.Date);
  const [sortOrder, setSortOrder] = useState(Order.ASC);
  const [sortedMeasurements, setSortedMeasurements] = useState(measurements);

  useEffect(() => {
    setSortedMeasurements(
      [...measurements].sort(
        columns.find((col) => col.id === sortedBy)?.sortFunction(sortOrder)
      )
    );
  }, [sortedBy, sortOrder, measurements]);

  const renderSortIcon = (col: Col) => {
    if (sortedBy !== col.id) {
      return;
    }

    if (sortOrder === Order.ASC) {
      return " \u2193";
    } else {
      return " \u2191";
    }
  };

  const changeOrder = (col: Col) => {
    if (sortedBy !== col.id) {
      setSortedBy(col.id);
      setSortOrder(Order.ASC);
    } else {
      setSortOrder(sortOrder * -1);
    }
  };

  return (
    <table className="w-full mx-10 text-center">
      <thead>
        <tr>
          {columns.map((col) => (
            <th
              className="w-1/2 cursor-pointer"
              onClick={() => changeOrder(col)}
              key={col.label}
            >
              {col.label}
              {renderSortIcon(col)}
            </th>
          ))}
        </tr>
      </thead>
      <tbody>
        {sortedMeasurements.map((measurement) => (
          <tr key={measurement.date}>
            <td>{measurement.date}</td>
            <td>{measurement.value} kg</td>
            {/*  // TODO: Get unit from data */}
          </tr>
        ))}
      </tbody>
    </table>
  );
};

export default MeasurementList;
