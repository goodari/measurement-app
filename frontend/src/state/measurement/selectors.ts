import { createSelector } from "@reduxjs/toolkit";
import { RootState } from "../configureStore";
import { Measurement, UIError } from "./types";
import { sortByDate, Order } from "../../utils/measurement";

export const measurementsSelector = (state: RootState): Measurement[] =>
  state.measurement.measurements;

export const latestMeasurementSelector = createSelector(
  measurementsSelector,
  (measurements) => {
    const sorted = [...measurements].sort(sortByDate(Order.DESC));
    return sorted[0];
  }
);

export const measurementSumSelector = createSelector(
  measurementsSelector,
  (measurements) => {
    return measurements.reduce(
      (sum, measurement) => (sum += measurement.value),
      0
    );
  }
);

export const measurementsLoadingSelector = (state: RootState): boolean =>
  state.measurement.loading;

export const measurementErrorSelector = (
  state: RootState
): UIError | undefined => state.measurement.error;
