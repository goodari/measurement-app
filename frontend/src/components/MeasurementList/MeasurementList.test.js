import renderer from "react-test-renderer";
import { mockMeasurements } from "../../test/measurement";
import MeasurementList from "./index";

it("renders correctly", () => {
  const tree = renderer.create(
    <MeasurementList measurements={mockMeasurements} />
  );
  expect(tree).toMatchSnapshot();
});
