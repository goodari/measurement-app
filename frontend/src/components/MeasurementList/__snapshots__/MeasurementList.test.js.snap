// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<table
  className="w-full mx-10 text-center"
>
  <thead>
    <tr>
      <th
        className="w-1/2 cursor-pointer"
        onClick={[Function]}
      >
        Date
         â†“
      </th>
      <th
        className="w-1/2 cursor-pointer"
        onClick={[Function]}
      >
        Value
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        2021-08-20T07:08:59.028Z
      </td>
      <td>
        2000
         kg
      </td>
    </tr>
    <tr>
      <td>
        2021-07-22T17:08:59.028Z
      </td>
      <td>
        2345
         kg
      </td>
    </tr>
    <tr>
      <td>
        2021-06-01T11:08:59.028Z
      </td>
      <td>
        321
         kg
      </td>
    </tr>
  </tbody>
</table>
`;
